core/ast_node.js

```diff
@@ -0,0 +1,656 @@
+ * @param {Blockly.Block} block The block that we want to find the top
+ *     connection on.
+ * @return {?Blockly.ASTNode} The astnode containing the top connection.
+ * @private
+ */
+Blockly.ASTNode.prototype.findTopASTConnection_ = function(block) {
+  var previousConnection = block.previousConnection;
+  var outputConnection = block.outputConnection;
+  var newAstNode;
+  if (previousConnection) {
+    newAstNode = Blockly.ASTNode.createConnectionNode(previousConnection);
+  } else if (outputConnection) {
+    newAstNode = Blockly.ASTNode.createConnectionNode(outputConnection);
+  }
+  return newAstNode;
+};
+
+
+/**
+ * Given a location in a stack of blocks find the next out connection. If the
+ * location is nested the next out location should be the connected input.
+ * @param {!Blockly.Block} block The source block for the current location.
+ * @return {?Blockly.ASTNode} The ast node holding thenext out connection or
+ *     block.
+ * @private
+ */
+Blockly.ASTNode.prototype.findOutLocationForBlock_ = function(block) {
+  var newLocation, newAstNode;
+
+  newLocation = block.previousConnection.targetConnection;
+  newAstNode = Blockly.ASTNode.createConnectionNode(newLocation);
+  if (!newLocation) {
+    newLocation = block.previousConnection;
+    newAstNode = Blockly.ASTNode.createStackNode(block);
+  }
+  return newAstNode;
+};
+
+/**
+ * Find the element to the right of the current element in the ast.
+ * @return {?Blockly.ASTNode} An ast node that wraps the next field, connection,
+ *     block, or workspace. Return null if there is no value to the right.
+ */
+Blockly.ASTNode.prototype.next = function() {
+  var newAstNode = null;
+
+  switch (this.type_) {
+    case Blockly.ASTNode.types.WORKSPACE:
+      var newX = this.wsCoordinate_.x + Blockly.ASTNode.wsMove_;
+      var newWsCoordinate = new goog.math.Coordinate(newX, this.wsCoordinate_.y);
+      newAstNode = Blockly.ASTNode.createWorkspaceNode(this.location_,
+          newWsCoordinate);
+      break;
+
+    case Blockly.ASTNode.types.STACK:
+      var nextTopBlock = this.navigateBetweenStacks_(true);
+      newAstNode = Blockly.ASTNode.createStackNode(nextTopBlock);
+      break;
+
+    case Blockly.ASTNode.types.OUTPUT:
+      newAstNode = Blockly.ASTNode.createBlockNode(this.location_.sourceBlock_);
+      break;
+
+    case Blockly.ASTNode.types.FIELD:
+      var parentInput = this.location_.getParentInput();
+      newAstNode = this.findNextForField_(this.location_, parentInput);
+      break;
+
+    case Blockly.ASTNode.types.INPUT:
+      var inputConnection = this.location_.connection;
+      newAstNode = this.findNextForInput_(inputConnection, this.location_);
+      break;
+
+    case Blockly.ASTNode.types.BLOCK:
+      var nextConnection = this.location_.nextConnection;
+      newAstNode = Blockly.ASTNode.createConnectionNode(nextConnection);
+      break;
+
+    case Blockly.ASTNode.types.PREVIOUS:
+      var output = this.location_.outputConnection;
```

Comment:
This is a previous, so there will never be an output. Also, this case should be identical to OUTPUT since in both cases next should go to the block itself.